--- a/usr/share/lipstick-jolla-home-qt5/compositor.qml
+++ b/usr/share/lipstick-jolla-home-qt5/compositor.qml
@@ -524,6 +524,7 @@ Compositor {
                     root._homeLayerOnPeek = homeLayerItem.currentItem
                     root.peekGestureStarted(root._peekDirection)
                     if (peekFilter.leftActive && Desktop.settings.left_peek_to_events) {
+                        switcherLayer.queuedWindow = topmostWindowId
                         homeLayerItem.setCurrentItem(homeLayerItem.events, !childrenOpaque, root._peekDirection)
                     } else {
                         homeLayerItem.setCurrentItem(homeLayerItem.switcher, !childrenOpaque, root._peekDirection)
--- a/usr/share/lipstick-jolla-home-qt5/layers/SwitcherLayer.qml
+++ b/usr/share/lipstick-jolla-home-qt5/layers/SwitcherLayer.qml
@@ -20,4 +20,17 @@ PannableLayer {
         id: overlay
         anchors.fill: parent
     }
+
+    property int queuedWindow: -1
+    onActiveChanged: {
+        /* Show queued window */
+        if (active == true && queuedWindow > -1) { // FIXME: lousy check
+            var windowItem = desktop.switcher.getItemForWid(queuedWindow)
+            if (windowItem != null) {
+                windowItem.open()
+            }
+            queuedWindow = -1
+        }
+    }
+
 }
--- a/usr/share/lipstick-jolla-home-qt5/switcher/SwitcherItem.qml
+++ b/usr/share/lipstick-jolla-home-qt5/switcher/SwitcherItem.qml
@@ -123,6 +123,14 @@ EditableGridDelegate {
         }
     }
 
+    function open() {
+        /*
+         * Opens the window.
+        */
+
+        clicked()
+    }
+
     function close() {
         _persistCloseApplicationEnabled |= switcherRoot.closeApplicationEnabled
         closeAnimation.start()
--- a/usr/share/lipstick-jolla-home-qt5/switcher/Switcher.qml
+++ b/usr/share/lipstick-jolla-home-qt5/switcher/Switcher.qml
@@ -95,6 +95,20 @@ SilicaFlickable {
         return false
     }
 
+    function getItemForWid(wid) {
+        /*
+         * Gets an item given a wid.
+        */
+
+        for (var i = 0; i < repeater.count; ++i) {
+            var item = repeater.itemAt(i)
+            if (item.windowId == wid) {
+                return item
+            }
+        }
+        return null
+    }
+
     property bool closePending
     property bool housekeepingMenuActive: housekeepingMenu.active
     onHousekeepingMenuActiveChanged: {
--- a/usr/share/lipstick-jolla-home-qt5/notifications/NotificationItem.qml
+++ b/usr/share/lipstick-jolla-home-qt5/notifications/NotificationItem.qml
@@ -189,6 +189,7 @@ Item {
             if (!notificationGroup || !notificationGroup.members.length) {
                 return
             }
+            Lipstick.compositor.switcherLayer.queuedWindow = -1
             invokeAppAction()
         }
     }
@@ -330,6 +331,7 @@ Item {
                             Lipstick.compositor.eventsLayer.housekeeping = false
                             return
                         }
+                        Lipstick.compositor.switcherLayer.queuedWindow = -1
                         notification.actionInvoked("default")
                     }
 
